calyx-opt/src/passes/static_inliner.rs

```diff
@@ -0,0 +1,388 @@
+use crate::analysis::ControlId;
+use crate::traversal::{Action, Named, VisResult, Visitor};
+use calyx_ir as ir;
+use calyx_ir::structure;
+use calyx_ir::LibrarySignatures;
+use ir::build_assignments;
+use std::collections::HashMap;
+use std::rc::Rc;
+
+#[derive(Default)]
+pub struct StaticInliner {
+    /// maps ids of static control stmts to groups
+    map: HashMap<u64, ir::RRC<ir::StaticGroup>>,
+}
+
+impl Named for StaticInliner {
+    fn name() -> &'static str {
+        "static-inline"
+    }
+
+    fn description() -> &'static str {
+        "Compiles Static Control into a single Static Enable"
+    }
+}
+
+impl StaticInliner {
+    // Updates the assignments so that they have appropriate timing in the
+    // inlined static seq group
+    // adds offset to each timing guard in `assigns`
+    // e.g., %[2,3] with offset = 2 -> %[4,5]
+    // all guards also must update so that guard -> guard & %[offset, offset+latency] since that
+    // is when the group will be active in the control, i.e., dst = guard ? src
+    // becomes dst =  guard & %[offset, offset+latency] ? src
+    fn update_assignment_timing(
+        assigns: &mut Vec<ir::Assignment<ir::StaticTiming>>,
+        offset: u64,
+        latency: u64,
+    ) {
+        for assign in assigns {
+            // adding the offset to each timing interval
+            assign.for_each_interval(|timing_interval| {
+                let (beg, end) = timing_interval.get_interval();
+                Some(ir::Guard::Info(ir::StaticTiming::new((
+                    beg + offset,
+                    end + offset,
+                ))))
+            });
+            // adding the interval %[offset, offset + latency]
+            assign.guard.add_interval(ir::StaticTiming::new((
+                offset,
+                offset + latency,
+            )));
+        }
+    }
+
+    // Makes assignments such that if branches can start executing on the first
+    // possible cycle.
+    // essentially, on the first cycle, we write port's value into a `cond` = a register.
+    // this is because the tru/false branch might alter port's value when it executes
+    // cond_wire reads from port on the first cycle, and then cond for the other cycles.
+    // this means that all of the tru branch assigns can get the cond_wire ? in front of them,
+    // and all false branch assigns can get !cond_wire ? in front of them
+    // makes the following assignments:
+    // read more here: https://github.com/cucapra/calyx/issues/1344 (specifically
+    // the section "Conditionl")
+    fn make_cond_assigns(
+        cond: &ir::RRC<ir::Cell>,
+        cond_wire: &ir::RRC<ir::Cell>,
+        port: &ir::RRC<ir::Port>,
```

Comment:
I generally avoid passing `RRC`'s by reference because they themselves represent a reference to something. Not sure if this has any profound impact on performance but more of a stylistic thing